apps/src/gamelab/GameLab.js

```diff
@@ -415,40 +427,73 @@ GameLab.prototype.onPuzzleComplete = function (submit) {
+    }
+    var containedResult = results[0].result;
+    this.testResults = results[0].testResult ? results[0].testResult :
+        containedResult.result ? this.studioApp_.TestResults.ALL_PASS :
+          this.studioApp_.TestResults.GENERIC_FAIL;
+    containedLevelInfo = {
+      app: results[0].app,
+      level: results[0].id,
+      callback: results[0].callback,
+      result: containedResult.result,
+      program: containedResult.response,
+    };
+  } else {
+    if (this.level.editCode) {
+      // If we want to "normalize" the JavaScript to avoid proliferation of nearly
+      // identical versions of the code on the service, we could do either of these:
 
-    // do an acorn.parse and then use escodegen to generate back a "clean" version
-    // or minify (uglifyjs) and that or js-beautify to restore a "clean" version
+      // do an acorn.parse and then use escodegen to generate back a "clean" version
+      // or minify (uglifyjs) and that or js-beautify to restore a "clean" version
 
-    program = this.studioApp_.getCode();
+      program = encodeURIComponent(this.studioApp_.getCode());
+    } else {
+      var xml = Blockly.Xml.blockSpaceToDom(Blockly.mainBlockSpace);
+      program = encodeURIComponent(Blockly.Xml.domToText(xml));
+    }
+  }
+
+  if (this.testResults >= this.studioApp_.TestResults.FREE_PLAY) {
+    this.studioApp_.playAudio('win');
   } else {
-    var xml = Blockly.Xml.blockSpaceToDom(Blockly.mainBlockSpace);
-    program = Blockly.Xml.domToText(xml);
+    this.studioApp_.playAudio('failure');
   }
 
   this.waitingForReport = true;
 
   var sendReport = function () {
-    this.studioApp_.report({
-      app: 'gamelab',
-      level: this.level.id,
-      result: levelComplete,
-      testResult: this.testResults,
-      submitted: submit,
-      program: encodeURIComponent(program),
-      image: this.encodedFeedbackImage,
-      onComplete: (submit ? this.onSubmitComplete.bind(this) : this.onReportComplete.bind(this))
-    });
+    if (containedLevelInfo) {
+      this.studioApp_.onAttempt({
+        callback: containedLevelInfo.callback,
+        app: containedLevelInfo.app,
+        level: containedLevelInfo.level,
+        result: containedLevelInfo.result,
+        testResult: this.testResults,
+        submitted: submit,
+        program: containedLevelInfo.program,
+        image: this.encodedFeedbackImage,
+        onComplete: (submit ? this.onSubmitComplete.bind(this) : this.onReportComplete.bind(this))
+      });
+    } else {
+      this.studioApp_.report({
+        app: containedLevelInfo ? containedLevelInfo.app : 'gamelab',
+        level: containedLevelInfo ? containedLevelInfo.level : this.level.id,
+        result: levelComplete,
+        testResult: this.testResults,
+        submitted: submit,
+        program: program,
+        image: this.encodedFeedbackImage,
+        onComplete: (submit ? this.onSubmitComplete.bind(this) : this.onReportComplete.bind(this))
+      });
+    }
```

Comment:
that is the plan, per my comments above. this is just a checkpoint since the merging of maintaining a parallel branch is costing me a lot of time every day.